### PR TITLE
feat: check that the images an install pulls match the compose it runs

### PR DESCRIPTION
**Pull request by goldyfruit**
Closes the fifth coupling in the cross-repo contract — the one that was declared and never checked.

## The gap

`contract.yml` has listed `images` since the contract landed, and `check_contracts.py` never read it. Worse, its docstring claimed `--online` verified them. That claim is corrected in its own commit at the head of this branch, before the code that makes it true.

Compose files come from a pinned git tag; images come from a **moving** channel tag. They can disagree, and when they do the install fails somewhere far from the cause — which is exactly how the `/home/ovos` report that started this work behaved.

## Why not "revision >= pinned ref"

Because it would be red every week until someone turned it off. An image is rebuilt whenever the channel's constraints move, with **no commit in the producer at all**, so its revision sits behind the tag permanently and legitimately. Today `ovos-core:testing` is 3 commits behind `v2.1.0` and entirely fine.

So the question asked is narrower: *between the image's revision and the pinned ref, did anything change that would have rebuilt **this** image?*

That already has an authoritative answer — the producer's own `scripts/affected.py`, the selector its CI uses to decide what to rebuild, run in a blobless clone at the pinned ref. Reusing it means **this check and the publisher cannot hold different opinions** about what affects an image. On today's lag it returns `[]`, so the harmless case is quiet by construction rather than by a tuned exception list:

```
REAL LAG (13a248cc..v2.1.0)  -> targets=[]          # quiet
v2.0.2..v2.1.0               -> 31 targets          # docker-bake.hcl changed
bogus revision               -> cannot tell         # never "clean"
```

## It finds a real one on the first run

```
images checked: hivemind-docker@testing 2/2
images checked: ovos-docker@testing 28/28

hivemind-cli:testing / hivemind-listener:testing / hivemind-satellite:testing
  built from 6087d744, which predates v2.1.1 by changes that rebuild it
```

Verified independently: `base/Dockerfile` changed between `6087d744` and `v2.1.1` (a Renovate uv bump), and `base/` is the root of that repository's bake graph, so every image is rebuilt by it. **ovos-docker reports nothing** — its lag is the harmless one.

## Design notes

- **Ownership routing.** Images are attributed via `org.opencontainers.image.source`, because ovos-docker's compose deploys `hivemind-cli` and `hivemind-listener`, which hivemind-docker builds. Judging those against ovos-docker's pin would be nonsense.
- **The ahead direction** compares the service definition the image must agree with — environment *keys*, volume targets, entrypoint — ignoring values, which are the installer's to set.
- **GHCR, not Docker Hub.** Installs pull from Docker Hub, but its unauthenticated budget is ~100 manifest requests/hour/IP shared across a runner's egress; using it here would break *other* jobs. GHCR's anonymous token API has no such limit.
- **Attestation manifests** in a multi-arch index have a config blob of literally `{}` — walking every entry would report "no revision label" for every image. Only real platforms are read.
- **No silent success.** Coverage is printed unconditionally, an unreachable registry is a note naming the reason, and an image with no bake target is *cannot tell*, never *clean*.

Needs **no contract change and no new producer release** — it reads the compose files and bake graph already present at the pinned tags.

## Tests

`scripts/test_image_coherence.py` — 11 offline tests against a throwaway git repository, so PR CI needs no registry. Both directions asserted for every guard, including the one that matters most: *an image behind the pin by a harmless commit is coherent*. Also that a missing bake target, a missing bake graph and an unreachable revision are each **unknown rather than clean**, and that changing an environment *value* is not an interface change while adding a *key* is.

172/172 bats, `ansible-lint` clean, syntax verified on 3.11 (the interpreter the virtualenv scenarios use) as well as 3.14.

Escalation is behind `--fail-on-image-drift`, deliberately separate from `--fail-on-stale`: it can go red for reasons a pin bump cannot fix, and conflating them would make one alert unreadable.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Contract verification now detects when container images have drifted from their expected build targets.
  * Image checks distinguish actionable mismatches from unresolved conditions such as missing tags, access restrictions, or registry failures.
  * Verification results now report image coverage counts, preventing runs with no verified images from appearing successful.

* **Tests**
  * Added coverage for image consistency checks and related edge cases.

### CODE REVIEW

File: scripts/image_coherence.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+        declared = set(contract.get("images") or [])
+        for image in sorted(set(deployed) - declared):
+            problems.append(f"{repo}: the pinned compose deploys {image} but the contract does not "
+                            f"declare it - the producer's own contract gate missed it")
+
+        resolved = 0
+        for image in sorted(deployed):
+            labels, outcome, used_repo = None, "no_tag", None
+            for candidate_repo, path in mirror_candidates(image, facts["slugs"], repo):
+                labels, outcome = read_labels(path, channel)
```
[the remark this answers is not in the document]
Comment (goldyfruit) [reply scripts/image_coherence.py:278]: Correct, and unaddressed when this merged. An adversarial audit reached the same conclusion independently.

`check_images` now picks the most severe outcome across candidates rather than letting the last one decide: `next((o for o in ("budget", "transport", "denied", "no_tag") if o in attempts), "no_tag")`. A registry that could not be reached is no longer reported as "no such tag" because a later mirror happened to 404.

Fixed in #606, with a test that drives two candidates returning different outcomes and asserts the first one wins.

File: scripts/image_coherence.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+                notes.append(f"{repo}: {image} is built by {source or 'an unknown repository'}, "
+                             f"which this installer does not pin - out of scope")
+                continue
+            owner_clone, owner_pin = clones.get(owner), facts["pins"].get(owner)
+            if owner_clone is None:
+                notes.append(f"{repo}: {image} is built by {owner}, which could not be cloned - "
+                             f"not verified")
+                continue
+
+            resolved += 1
```
[the remark this answers is not in the document]
Comment (goldyfruit) [reply scripts/image_coherence.py:309]: Correct, and unaddressed when this merged. The audit found the same thing and rated it the most serious of the ten: a run could reach no verdict about any image and still exit 0, because the coverage figure was printed but never compared against itself.

`resolved` is now incremented only on a decisive verdict, and `check_contracts.main` asserts `resolved == total` rather than merely printing it — an image that reached no verdict is a violation under `--fail-on-image-drift`. The count is also seeded before every early exit, so a repository that drops out has a `0/0 (not checked)` line instead of no line at all.

Fixed in #606.

File: scripts/image_coherence.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+        return ("behind", f"rebuilds {target}") if target in targets else ("coherent", "unaffected")
+
+    # Ahead: the image moved past the compose. What matters is whether the part of the service
+    # definition the image has to agree with changed underneath it.
+    changed = []
+    for compose_name, service in (placements or {}).items():
+        at_pin = service_interface(compose_clone, compose_pin, compose_name, service)
+        at_rev = service_interface(clone, revision, compose_name, service)
+        if at_pin is None or at_rev is None:
+            continue
```
[the remark this answers is not in the document]
Comment (goldyfruit) [reply scripts/image_coherence.py:368]: Correct, and unaddressed when this merged. The audit reached the same conclusion, and found a second cause of the same symptom alongside it.

`compare` now collects unreadable placements and returns `unknown` unless another placement proves drift — absence of evidence is not evidence of agreement. The second cause: the service definition was being read out of the *builder's* tree rather than the *deployer's*, so for every cross-repository image (ovos-docker's compose deploys `hivemind-cli`, and hivemind-docker has no such file) the lookup came back empty and read as agreement.

Fixed in #606, with a test for each cause.
